File: yearEnd2018/july2017/detect_capital.cpp
// Problem url: https://leetcode.com/problems/detect-capital/description/

#include <cctype>

using namespace std;

class Solution {
public:
    bool detectCapitalUse(string word) {
        // Base case, avoiding crashes 
        if(word.length() == 0 ) 
            return true;
        
        bool is_first_big = isupper(word[0]);
        int counter = 1;
        
        for(int i=1; i<word.size(); i++) {
            bool b = isupper(word[i]);
            
            // If any big letter is found
            if( b == is_first_big )
                counter++;
        }
        
        // This is for the first letter big
        if( is_first_big && counter == 1 )
            return true;
        
        // This is for all small of all big
        if( counter == word.size() )
            return true;
        else 
            return false;
    }
};

File: Year2020/leetcode20/src/main/java/alpha/DetectCapital.java
/*
* Problem URL: https://leetcode.com/problems/detect-capital/
* Problem ID: 520
* Status: Accepted
* Difficulty: Easy
*/

package alpha;

public class DetectCapital {
    private String wordToBeChecked;

    private boolean performCheck(boolean isFirstLetterBig) {
        boolean retValue = true;
        boolean allSmallIsRequired = false;

        // Starting from 1 as the first letter is already checked
        for( int i=1; i < wordToBeChecked.length(); i++ ) {
            char testChar = wordToBeChecked.charAt(i);

            // This mean all the letter should be small
            if( !isFirstLetterBig ) {
                if( Character.isUpperCase(testChar) ) {
                    retValue = false;
                    break;
                }
            }
            else {
                if( i == 1 && !Character.isUpperCase(testChar)) {
                    allSmallIsRequired = true;
                    continue;
                }

                boolean t = !Character.isUpperCase(testChar);
                if( t != allSmallIsRequired ) {
                    retValue = false;
                    break;
                }
            }
        }

        return retValue;
    }

    public boolean detectCapitalUse(String word) {
        // Sanity checking
        if( word == null ) return false;

        // A single length string would always be fine
        int wordLength = word.length();
        if( wordLength == 1 ) return true;

        // Setting up for the main function call
        char firstChar = word.charAt(0);
        boolean isUpper = Character.isUpperCase(firstChar);
        wordToBeChecked = word;

        // The method which would do the actual check
        return performCheck(isUpper);
    }
}
